src/compiler/scala/tools/reflect/StringInterpolator.scala

```diff
@@ -0,0 +1,162 @@
+package scala.tools.reflect
+
+import scala.annotation.tailrec
+import scala.reflect.macros.runtime.Context
+
+abstract class StringInterpolator {
+  val c: Context
+  val global: c.universe.type = c.universe
+
+  import c.universe.{Match => _, _}
+  import definitions._
+  import treeInfo.Applied
+
+  @inline private def truly(body: => Unit): Boolean = {
+    body
+    true
+  }
+
+  @inline private def falsely(body: => Unit): Boolean = {
+    body
+    false
+  }
+
+  private def bail(msg: String) = global.abort(msg)
+
+  def interpolateString: Tree = interpolateImpl(false)
+
+  def interpolateRaw: Tree = interpolateImpl(true)
+
+  def interpolateImpl(raw: Boolean): Tree = c.macroApplication match {
+    case Applied(Select(Apply(_, parts), _), _, argss) =>
+      val args = argss.flatten
+
+      def badlyInvoked = (parts.length != args.length + 1) && truly {
+        //probably cant use s"..." here as this is the macro for s
+        def because(s: String) = "too " + s + " arguments for interpolated string"
+
+        val (p, msg) =
+          if (args.length + 1 < parts.length)
+            (if (args.isEmpty) c.enclosingPosition else args.last.pos, because("few"))
+          else (args(parts.length - 1).pos, because("many"))
+        c.abort(p, msg)
+      }
+
+      if (badlyInvoked) c.macroApplication else {
+        val res = interpolated(raw, parts, args)
+        /*if (settings.debug) */reporter.echo(c.enclosingPosition, "rewrite expression to "+res)
+        res
+      }
+    case other =>
+      bail("Unexpected application " + showRaw(other))
+      other
+  }
+
+  def interpolated(raw: Boolean, parts: List[Tree], args: List[Tree]): Tree = {
+    var literalSize = 0
+    val literals: List[List[Literal]] = parts map {
+      case lit@Literal(Constant(x: String)) =>
+        if (x.isEmpty) Nil
+        else {
+          literalSize += x.length
+          if (raw) lit :: Nil
+          else (Literal(Constant(/*StringContext.*/ treatEscapes(x))) setPos (lit.pos)) :: Nil
+        }
+      case _ => throw new IllegalArgumentException("internal error: argument parts must be a list of string literals")
+    }
+    if (args isEmpty)
+    //it of the form s"some text" or s""
+      parts(0)
+    else if (args.tail.isEmpty && literalSize == 0)
+    //it of the form s"$value", so we dont need a StringBuilder
+      Apply(Select(Ident(StringModule), termNames.valueOf), args.head :: Nil) setPos (args.head.pos)
```

Comment:
This needs to select the `valueOf(Object)String` overload specfically. Right now, it changes behaviour:

```
scala> s"${new Array[Char](0)}"
res0: String = [C@119d4443

scala> sNew"${new Array[Char](0)}"
<console>:12: rewrite expression to String.valueOf(new Array[Char](0))
       sNew"${new Array[Char](0)}"
       ^
res1: String = ""
```